uui-core/src/data/processing/views/tree/treeStructure/helpers/PatchHelper.ts

```diff
@@ -0,0 +1,192 @@
+import { IImmutableMap, IMap } from '../../../../../../types';
+import { numberToOrder } from '../../../../../../helpers';
+import { buildComparators, composeComparetors } from '../../helpers';
+import { NOT_FOUND_RECORD } from '../../exposed';
+import { ItemsAccessor } from '../../ItemsAccessor';
+import { TreeStructure } from '../TreeStructure';
+import { cloneMap } from '../../helpers/map';
+import { merge } from '../../helpers/merge';
+import { PatchIntoTreeStructureOptions } from './types';
+import { ItemsMap } from '../../ItemsMap';
+
+interface ApplyPatchWithSortingOptions<TItem, TId> {
+    comparator: (a: TItem, b: TItem) => number;
+    patchAtLastSort: IMap<TId, TItem> | IImmutableMap<TId, TItem>;
+    originalItemsMap: ItemsMap<TId, TItem>;
+    patchedItemsMap: ItemsMap<TId, TItem>;
+    isDeleted: (id: TId) => boolean;
+    complexIds?: boolean;
+}
+
+interface ApplyPatchTemporaryReorderingOptions<TItem, TId> {
+    getItemTemporaryOrder: (item: TItem) => string;
+    patchedItemsMap: ItemsMap<TId, TItem>;
+    isDeleted: (id: TId) => boolean;
+    complexIds?: boolean;
+}
+
+export class PatchHelper {
+    private static applyPatchWithSorting<TItem, TId>(
+        patchIds: TId[],
+        originalIds: TId[],
+        {
+            comparator,
+            patchAtLastSort,
+            originalItemsMap,
+            patchedItemsMap,
+            isDeleted,
+            complexIds,
+        }: ApplyPatchWithSortingOptions<TItem, TId>,
+        initialIds: TId[] = [],
+    ) {
+        return merge(
+            patchIds,
+            originalIds,
+            (patchItemId, itemId) => {
+                const inPatchBeforeSort = patchAtLastSort.has(patchItemId);
+                const inOriginalTree = originalItemsMap.has(patchItemId);
+
+                let patchItemToCompare;
+                if (inPatchBeforeSort) {
+                    patchItemToCompare = patchAtLastSort.get(patchItemId);
+                } else if (inOriginalTree) {
+                    patchItemToCompare = originalItemsMap.get(patchItemId);
+                } else {
+                    patchItemToCompare = patchedItemsMap.get(patchItemId);
+                }
+
+                const item = originalItemsMap.get(itemId);
+                return comparator(patchItemToCompare, item);
+            },
+            initialIds,
+            { isDeleted, complexIds },
+        );
+    }
+
+    private static applyPatchTemporaryReordering<TItem, TId>(
+        patchIds: TId[],
+        originalIds: TId[],
+        {
+            getItemTemporaryOrder,
+            patchedItemsMap,
+            isDeleted,
+            complexIds,
+        }: ApplyPatchTemporaryReorderingOptions<TItem, TId>,
+    ) {
+        const tempOrderComparator = new Intl.Collator(undefined, { numeric: true, sensitivity: 'base' }).compare;
```

Comment:
We should use a simple, non Intl sting comparator here, to compare orders. It's faster and don't have issues with digits.

Reply:
Done.